staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -156,11 +174,63 @@ type crdInfo struct {
 	storageVersion string
 
 	waitGroup *utilwaitgroup.SafeWaitGroup
+
+	// storageVersionUpdate holds information about the storage version
+	// update issued when the crdInfo was created. The API server uses the
+	// information to decide whether a CR write request should be allowed,
+	// rejected, or blocked.
+	storageVersionUpdate *storageVersionUpdate
 }
 
 // crdStorageMap goes from customresourcedefinition to its storage
 type crdStorageMap map[types.UID]*crdInfo
 
+// storageVersionUpdate holds information about a storage version update,
+// indicating whether the update gets processed, or timed-out.
+type storageVersionUpdate struct {
+	// processedCh is closed by the storage version manager after the
+	// storage version update gets processed (either succeeded or failed).
+	// The API server will unblock and allow CR write requests if this
+	// channel is closed.
+	processedCh <-chan struct{}
+
+	// timeout is the time when the API server will unblock and allow CR
+	// write requests even if the storage version update hasn't been
+	// processed.
+	timeout time.Time
+}
+
+func (i *crdInfo) waitForStorageVersionUpdate(ctx context.Context) error {
+	// StorageVersionAPI feature gate is disabled
+	if !utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		klog.V(2).Infof("Skipped waiting for storage version to finish updating since StorageVersionAPI and/or APIServerIdentity feature are disabled.")
+		return nil
+	}
+
+	if i.storageVersionUpdate == nil {
+		return nil
+	}
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-i.storageVersionUpdate.processedCh:
+		return nil
```

Comment:
When the `UpdateStorageVersion`, how does the caller of this `waitForStorageVersionUpdate` know the write was unsuccessful?

Reply:
Added a new errCh in the storageVersionUpdate struct to signal that an error has occurred while waiting for the storageversion to be updated and hence the write to the CR should fail. How does this look?

Reply:
Also, is it correct to exit out on waitForStorageVersionUpdate() if UpdateStorageVersion had failed, which would cause all subsequent writes to a CR to fail?